transport/ssh.go

```diff
@@ -119,68 +117,61 @@ func ExecSSH(ctx context.Context, id Identity, cmd []string, opts plugin.ExecOpt
 
 	strictHostKeyChecking, err := ssh_config.GetStrict(id.Host, "StrictHostKeyChecking")
 	if err != nil {
-		return result, err
+		return nil, err
 	}
 
 	connection, err := sshConnect(id.Host, port, user, strictHostKeyChecking != "no")
 	if err != nil {
-		return result, fmt.Errorf("Failed to connect: %s", err)
+		return nil, fmt.Errorf("Failed to connect: %s", err)
 	}
 
 	// Run command via session
 	session, err := connection.NewSession()
 	if err != nil {
-		return result, fmt.Errorf("Failed to create session: %s", err)
+		return nil, fmt.Errorf("Failed to create session: %s", err)
 	}
 
 	if opts.Tty {
 		// sshd only processes signal codes if a TTY has been allocated. So set one up when requested.
 		modes := ssh.TerminalModes{ssh.ECHO: 0, ssh.TTY_OP_ISPEED: 14400, ssh.TTY_OP_OSPEED: 14400}
 		if err := session.RequestPty("xterm", 40, 80, modes); err != nil {
-			return result, fmt.Errorf("Unable to setup a TTY: %v", err)
+			return nil, fmt.Errorf("Unable to setup a TTY: %v", err)
 		}
 	}
 
-	outputch, stdout, stderr := plugin.CreateExecOutputStreams(ctx)
-	session.Stdin, session.Stdout, session.Stderr = opts.Stdin, stdout, stderr
+	cmdObj := plugin.NewRunningCommand(ctx)
+	session.Stdin, session.Stdout, session.Stderr = opts.Stdin, cmdObj.Stdout(), cmdObj.Stderr()
 
 	if opts.Elevate {
 		cmd = append([]string{"sudo"}, cmd...)
 	}
 
 	cmdStr := shellquote.Join(cmd...)
 	if err := session.Start(cmdStr); err != nil {
-		return result, err
+		return nil, err
 	}
-
-	// Wait for session to complete and stash result. This is done asynchronously because the
-	// exitcode callback won't be called until stdout/stderr are closed, but they're not closed by
-	// session.Start.
-	var exitErr error
-	go func() {
-		exitErr = session.Wait()
-		activity.Record(ctx, "Closing session for %v: %v", id.Host, session.Close())
-		stdout.Close()
-		stderr.Close()
-	}()
-
-	result.OutputCh = outputch
-	result.CancelFunc = func() {
+	cmdObj.SetStopFunc(func() {
```

Comment:
Moving this closer to `Start` makes me think of this as "Ok, the command's started so here we're setting the stop function". Can move back if it makes things more confusing.